### PR TITLE
Implementation of Subclasses of Widgets to enable multi zoom level support for win32

### PR DESCRIPTION
**Pull request by amartya4256**
## Addressed issues
* https://github.com/eclipse-platform/eclipse.platform.swt/issues/62 
* https://github.com/eclipse-platform/eclipse.platform.swt/issues/131

## Requires
* [x] https://github.com/eclipse-platform/eclipse.platform.swt/pull/1214

**Note:** Only the last commit in this PR is to be reviewed. Previous commit(s) belong to the prerequisite PR(s)

## Description

This pull request is based on the implementations of PR #1214. It uses the native zoom of widget to do the scaling using the DPIUtil methods with zoom level. The Event class also uses DPIUtil in a few methods which are used by the win32 widget classes. This contribution adds the reference of the zoom level fom the widget from those calls in these methods within Event.

The test assertion is purposely commented out since the access to the widget zoom is no longer available outside the package. We might want to create a test fragment for the package itself so that its easy to test package protected members.

### CODE REVIEW

**Comment by amartya4256** [comment]
@HeikoKlare @fedejeanne @akoch-yatta Please have a look.

File: tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/graphics/WidgetWin32Tests.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		Shell shell = new Shell(display);
+
+		Button button = new Button(shell, SWT.PUSH);
+		button.setBounds(0, 0, 200, 50);
+		button.setText("Widget Test");
+		button.setBackground(display.getSystemColor(SWT.COLOR_CYAN));
+		shell.open();
+		changeZoomLevel(scaledZoom, shell);
+		// TODO: Zoom level not accessible from this test, Would be fixed later.
+//		assertEquals("The Zoom Level should be updated for button on zoom change event on its shell", scaledZoom, button.getZoom());
```
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/graphics/WidgetWin32Tests.java:49]: Please fix this :-)

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Widget.java [hunk 2]
```diff
@@ -1232,6 +1232,7 @@ boolean sendMouseEvent (int type, int button, long hwnd, long lParam) {
 boolean sendMouseEvent (int type, int button, int count, int detail, boolean send, long hwnd, long lParam) {
 	if (!hooks (type) && !filters (type)) return true;
 	Event event = new Event ();
+	event.widget = this;
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Widget.java:1235]: This change seems to be unrelated to this PR. Is this some independent bug? Can we extract that change?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/TreeItem.java [hunk 3]
```diff
@@ -375,7 +375,7 @@ public Color getBackground (int index) {
  */
 public Rectangle getBounds () {
 	checkWidget ();
-	return DPIUtil.autoScaleDown(getBoundsInPixels());
+	return DPIUtil.scaleDown(getBoundsInPixels(), parent.getZoom());
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/TreeItem.java:378]: Why is `getZoom()` called on the parent at some places (e.g., `TreeItem`, `TreeColum`, `Tracker`, ...) instead of just calling `getZoom()` on the widget iself?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java [hunk 4]
```diff
@@ -7928,8 +7928,9 @@ LRESULT wmNotifyHeader (NMHDR hdr, long wParam, long lParam) {
 							GC gc = getNewGC(nmcd.hdc, data);
 							Rectangle imageBounds = DPIUtil.autoScaleBounds(columns[i].image.getBounds(), this.getZoom(), 100);
 							int y = Math.max (0, (nmcd.bottom - imageBounds.height) / 2);
-							gc.drawImage (columns[i].image, DPIUtil.autoScaleDown(x), DPIUtil.autoScaleDown(y));
-							x += imageBounds.width + 12;
+							Rectangle scaledBounds = DPIUtil.autoScaleBounds(columns[i].image.getBounds(), getZoom(), 100);
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java:7931]: I do not understand the necessity for the `scaledBounds`. Aren't they calculated in exactly the same way in which `imageBounds` is calculated?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java [hunk 5]
```diff
@@ -757,24 +757,24 @@ LRESULT CDDS_ITEMPOSTPAINT (NMTVCUSTOMDRAW nmcd, long wParam, long lParam) {
...
 							//int y1 = rect.top + (index == 0 ? (getItemHeight () - size.y) / 2 : 0);
-							int y1 = rect.top + DPIUtil.autoScaleUp((getItemHeight () - size.y) / 2);
+							int y1 = rect.top + DPIUtil.autoScaleUp((getItemHeight () - size.y) / 2, getZoom());
 							int x1 = Math.max (rect.left, rect.left - inset + 1);
 							GCData data = new GCData();
 							data.device = display;
-							GC gc = getNewGC(hDC, data);
-							gc.setClipping (DPIUtil.autoScaleDown(new Rectangle(x1, rect.top, rect.right - x1, rect.bottom - rect.top)));
-							gc.drawImage (image, 0, 0, bounds.width, bounds.height, DPIUtil.autoScaleDown(x1), DPIUtil.autoScaleDown(y1), size.x, size.y);
+							GC gc = GC.win32_new (hDC, data);
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java:771]: This seems to revert changes of a previous commit. Is that intended?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/TaskItem.java [hunk 6]
```diff
@@ -462,12 +462,9 @@ public void setProgressState (int progressState) {
...
 	this.shell = shell;
-	shell.addListener (SWT.Dispose, new Listener () {
-		@Override
-		public void handleEvent (Event event) {
-			if (isDisposed ()) return;
-			dispose ();
-		}
+	shell.addListener (SWT.Dispose, event -> {
+		if (isDisposed ()) return;
+		dispose ();
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/TaskItem.java:467]: Please revert this unrelated change.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Display.java [hunk 7]
```diff
@@ -3007,6 +3007,14 @@ Point mapInPixels (Control from, Control to, int x, int y) {
 	return new Point (point.x, point.y);
 }
 
+private int getZoomLevelForMapping(Control from, Control to) {
+	return Optional.ofNullable(to)
+            .map(Control::getZoom)
+            .orElseGet(() -> Optional.ofNullable(from)
+                    .map(Control::getZoom)
+                    .orElse(0));
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Display.java:3015]: Two question on this:
1. Does it make sense to default the zoom value to "0"? In case this really happens, the results will be completely unusable, so I assume you do not expect this case to occur at all. Wouldn't it be better to thrown an exception then to make explicit what goes wrong if that case really happens?
2. I am not sure whether using stream operations is a good idea here. It produces quite some overhead and the "simple" implementation would, in my opinion, be even easier to understand, as would look something like this, wouldn't it:
```java
if (to != null) {
	return to.getZoom();
}
return from.getZoom();
```

File: bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java [hunk 8]
```diff
@@ -357,13 +359,26 @@ public void setBounds (Rectangle rect) {
 	this.height = rect.height;
 }
 
+private int getDeviceZoom() {
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java:362]: This should just be `getZoom()`, shouldn't it?

File: bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java [hunk 9]
```diff
@@ -334,15 +334,17 @@ public Rectangle getBounds () {
 	return new Rectangle (x, y, width, height);
 }
 Rectangle getBoundsInPixels () {
-	return DPIUtil.autoScaleUp(getBounds());
+	int deviceZoom = getDeviceZoom();
+	return DPIUtil.autoScaleUp(getBounds(), deviceZoom);
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java:338]: You can simply inline the method call instead of assigning it a variable (as the method call is self-explaining) like you do in all the other classes.

File: bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java [hunk 10]
```diff
@@ -357,13 +359,26 @@ public void setBounds (Rectangle rect) {
...
 
+private int getDeviceZoom() {
+	if (widget != null) {
+		return widget.getZoom();
+	} else if (item != null) {
+		return item.getZoom();
+	} else if (gc != null && gc.getGCData() != null) {
+		return DPIUtil.getZoomForAutoscaleProperty(gc.getGCData().nativeDeviceZoom);
+	} else {
+		return DPIUtil.getDeviceZoom();
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java:370]: Do you know if/when this case can happen? Using `DPIUtil.getDeviceZoom()` should be something we only do for reasons of compatibility but not because it's actually needed, shouldn't it? I would expect the complete Windows SWT implementation to not use `DPIUtil.getDeviceZoom()` after HiDPI support has been completed at all.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java (hunk 4 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java:7931]: Yes you are right. There has been a mistake here.

File: bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java (hunk 10 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java]: Reverted changes for Event, since it is in common. @akoch-yatta has a separate PR for this.

File: bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java (hunk 9 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java]: Reverted changes for Event, since it is in common. @akoch-yatta has a separate PR for this.

File: bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java (hunk 8 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java]: Reverted changes for Event, since it is in common. @akoch-yatta has a separate PR for this.

File: bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java [hunk 11]
```diff
@@ -376,4 +376,4 @@ void setLocationInPixels (int x, int y) {
 public String toString () {
 	return "Event {type=" + type + " " + widget + " time=" + time + " data=" + data + " x=" + x + " y=" + y + " width=" + width + " height=" + height + " detail=" + detail + "}";  //$NON-NLS-1$//$NON-NLS-2$ //$NON-NLS-3$ //$NON-NLS-4$ //$NON-NLS-5$ //$NON-NLS-6$ //$NON-NLS-7$
 }
-}
+}
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java:379]: Please revert this unrelated/unnecessary change.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Display.java [hunk 12]
```diff
@@ -3008,6 +3008,13 @@ Point mapInPixels (Control from, Control to, int x, int y) {
 	return new Point (point.x, point.y);
 }
 
+private int getZoomLevelForMapping(Control from, Control to) {
+	if (to != null) {
+		return to.getZoom();
+	}
+	return from.getZoom();
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Display.java:3015]: At other places, the `from` and `to` values are checked and throw report proper SWT error, such as 
```java
if (from != null && from.isDisposed()) error (SWT.ERROR_INVALID_ARGUMENT);
```
Should this be done here as well?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java [hunk 13]
```diff
@@ -7928,8 +7928,8 @@ LRESULT wmNotifyHeader (NMHDR hdr, long wParam, long lParam) {
 							GC gc = createNewGC(nmcd.hdc, data);
 							Rectangle imageBounds = DPIUtil.autoScaleBounds(columns[i].image.getBounds(), this.getZoom(), 100);
 							int y = Math.max (0, (nmcd.bottom - imageBounds.height) / 2);
-							gc.drawImage (columns[i].image, DPIUtil.autoScaleDown(x), DPIUtil.autoScaleDown(y));
-							x += imageBounds.width + 12;
+							x += imageBounds.width + DPIUtil.autoScaleUp(12, getZoom());
+							gc.drawImage (columns[i].image, DPIUtil.scaleDown(x, getZoom()), DPIUtil.scaleDown(y, getZoom()));
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java:7932]: Is it inteded that the two statements have been swapped? That seems to modify the behavior.

File: tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/graphics/WidgetWin32Tests.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.swt.widgets.Event;
+import org.eclipse.swt.widgets.Shell;
+import org.eclipse.swt.widgets.WidgetTestUtil;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class WidgetWin32Tests {
```
Comment (HeikoKlare) [inline tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/graphics/WidgetWin32Tests.java:18]: Why is this class placed in the `graphics` instead of the `widgets` package? The `WidgetTestUtil` also seems to be a symptom of this potential misplacement.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Display.java (hunk 12 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Display.java:3015]: Yeah right. it is used in map methods which uses mapInPixels, so it could have the same effect.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java (hunk 13 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java:7932]: hmm seems like a mistake but just thinking if adding to x after using it even has any effect. Seems weird.

File: tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/graphics/WidgetWin32Tests.java (hunk 14 above)
Comment (amartya4256) [reply tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/graphics/WidgetWin32Tests.java:18]: Just wanted to have all these tests for multi zoom level in one package. But if its not a problem then I can move them to separate packages.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java (hunk 13 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java]: Yes I tested it. Seems consitent. You are right.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/CoolItem.java [hunk 15]
```diff
@@ -495,7 +495,7 @@ Point getSizeInPixels() {
 	rect.left -= margins.cxLeftWidth;
 	rect.right += margins.cxRightWidth;
 	if (!parent.isLastItemOfRow (index)) {
-		rect.right += (parent.style & SWT.FLAT) == 0 ? CoolBar.SEPARATOR_WIDTH : 0;
+		rect.right += (parent.style & SWT.FLAT) == 0 ? DPIUtil.autoScaleUp(CoolBar.SEPARATOR_WIDTH, getZoom())  : 0;
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/CoolItem.java:498]: This changes the existing behavior and rather seems like a fix for an existing bug than a necessity in terms of multi-zoom support (also see my other detailed comment on changed behavior). I would propose to address such existing bugs and the adaptation for multi-zoom support (of which the latter is the goal of this PR) in different PRs to separate concerns and ease review and discussion.

**Review by HeikoKlare — COMMENTED** [review]
My previous concern almost seem to be addressed, but the recent change seems to introduce faulty behavior (see my detailed comment).
In addition, there are some places where the scaling behavior of constants has been changed. I guess the calculation was erroneous before, but then those changes should be kept separate from this PR which is only supposed to introduce multi-zoom support and should not fix other bugs as a side effect.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Label.java [hunk 16]
```diff
@@ -137,9 +137,9 @@ static int checkStyle (int style) {
 	if ((style & SWT.SEPARATOR) != 0) {
 		int lineWidth = OS.GetSystemMetrics (OS.SM_CXBORDER);
 		if ((style & SWT.HORIZONTAL) != 0) {
-			width = DEFAULT_WIDTH;  height = lineWidth * 2;
+			width = DPIUtil.autoScaleUp(DEFAULT_WIDTH, getZoom());  height = lineWidth * 2;
```
[the remark this answers is not in the document]
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Label.java]: Basically I have removed the scaling from everywhere where a scaling was not applicable earlier.

File: tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/widgets/WidgetWin32Tests.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+		Button button = new Button(shell, SWT.PUSH);
+		button.setBounds(0, 0, 200, 50);
+		button.setText("Widget Test");
+		button.setBackground(display.getSystemColor(SWT.COLOR_CYAN));
+		shell.open();
+		changeZoomLevel(scaledZoom, shell);
+		assertEquals("The Zoom Level should be updated for button on zoom change event on its shell", scaledZoom, button.getZoom());
+	}
+
+	public void changeZoomLevel(int zoomLevel, Control widget) {
```
Comment (fedejeanne) [inline tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/widgets/WidgetWin32Tests.java:43]: Does this one really need to be `public`?

File: tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/widgets/WidgetWin32Tests.java [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+		display = Display.getDefault();
+	}
+
+	@After
+	public void tearDown() {
+		DPITestUtil.setAutoScaleOnRunTime(autoScaleOnRuntime);
+	}
+
+	@Test
+	public void widgetZoomShouldChangeOnZoomLevelChange() {
```
Comment (fedejeanne) [inline tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/widgets/WidgetWin32Tests.java:29]: Please name the test methods with the prefix `test`, it makes it easier to spot tests in the call hierarchy view.

**Review by fedejeanne — CHANGES_REQUESTED** [review]
@amartya4256 have Heiko's comments been addressed? I think they have, If they have been addressed then please explicitly answer in the threads and mark them as "resolved".

I noticed the tests are not being executed, can it be?

Also, please name the tests with the prefix "test" so one can spot them easier in the call hierarchy view.

File: tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/graphics/WidgetWin32Tests.java (hunk 1 above)
Comment (amartya4256) [reply tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/graphics/WidgetWin32Tests.java]: Done

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Widget.java (hunk 2 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Widget.java]: It was not needed anymore. Removed it.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/TreeItem.java (hunk 3 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/TreeItem.java]: Changed back to getZoom()

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java (hunk 5 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java]: Nope. Probably because of rebasing. Removed it.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/TaskItem.java (hunk 6 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/TaskItem.java]: Done

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Display.java (hunk 7 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Display.java]: Adapted as requested

File: bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java (hunk 11 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Event.java]: Done

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java [hunk 19]
```diff
@@ -7928,8 +7928,8 @@ LRESULT wmNotifyHeader (NMHDR hdr, long wParam, long lParam) {
 							GC gc = createNewGC(nmcd.hdc, data);
 							Rectangle imageBounds = DPIUtil.autoScaleBounds(columns[i].image.getBounds(), this.getZoom(), 100);
 							int y = Math.max (0, (nmcd.bottom - imageBounds.height) / 2);
-							gc.drawImage (columns[i].image, DPIUtil.autoScaleDown(x), DPIUtil.autoScaleDown(y));
-							x += imageBounds.width + 12;
+							gc.drawImage (columns[i].image, DPIUtil.scaleDown(x, getZoom()), DPIUtil.scaleDown(y, getZoom()));
+							x += imageBounds.width + DPIUtil.autoScaleUp(12, getZoom());
```
[the remark this answers is not in the document]
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Tree.java]: Reverted to the previous state. I think a separate PR is not needed for now.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/CoolItem.java (hunk 15 above)
Comment (amartya4256) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/CoolItem.java]: Removed DPIUTil calls for all the places where it wasnt there earlier

File: tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/widgets/WidgetWin32Tests.java (hunk 17 above)
Comment (amartya4256) [reply tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/widgets/WidgetWin32Tests.java]: nope. removed the whole method since it is not needed anymore. Extending to the base test for the method

File: tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/widgets/WidgetWin32Tests.java (hunk 18 above)
Comment (amartya4256) [reply tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/widgets/WidgetWin32Tests.java]: Done

File: tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/widgets/WidgetWin32Tests.java [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+import static org.junit.Assert.assertEquals;
+
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.internal.DPITestUtil;
+import org.eclipse.swt.internal.DPIUtil;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class WidgetWin32Tests {
```
[the remark this answers is not in the document]
Comment (amartya4256) [reply tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/widgets/WidgetWin32Tests.java]: Moved it to the other class where Andy had implemented similar tests in #1272

File: bundles/org.eclipse.swt/Eclipse SWT Tests/win32/org/eclipse/swt/widgets/WidgetWin32Tests.java [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+		int zoom = DPIUtil.getDeviceZoom();
+		int scaledZoom = zoom * 2;
+
+		Button button = new Button(shell, SWT.PUSH);
+		button.setBounds(0, 0, 200, 50);
+		button.setText("Widget Test");
+		button.setBackground(shell.getDisplay().getSystemColor(SWT.COLOR_CYAN));
+		shell.open();
+		changeDPIZoom(scaledZoom);
+		assertEquals("The Zoom Level should be updated for button on zoom change event on its shell", scaledZoom, button.getZoom());
```
Comment (fedejeanne) [inline bundles/org.eclipse.swt/Eclipse SWT Tests/win32/org/eclipse/swt/widgets/WidgetWin32Tests.java:22]: Nit: I would explicitly check for this to make the test clearer and more future-proof:

```suggestion
		assertEquals("The initial zoom is wrong", zoom, button.getZoom()); //pre-condition
		changeDPIZoom(scaledZoom);
		assertEquals("The Zoom Level should be updated for button on zoom change event on its shell", scaledZoom, button.getZoom());
```

**Review by fedejeanne — CHANGES_REQUESTED** [review]
Thank you for addressing the comments @amartya4256 , it's looking way better now.

I just have one Nit-Picky comment about the test (with code proposal) and a proposal regarding the commit text (hence the request for changes)

I see in the body of the commit text that:

```
This contribution removes the dependency to the static deviceZoom in DPIUtil for all widgets in the win32 implementation. All calls to DPIUtil method are replaced with calls passing the zoom of the widget directly.
```

How about these calls still in the code? I assume the one in `Display` is still OK (please confirm/rebut) but shouldn't the other ones be removed (in future PRs)?

image

Proposal: reword the commit to:
```
Multi zoom level for Widgets for win32

Do not use DPIUtil::getDeviceZoom to scale the widgets, use DPIUtil::autoScaleUp/scaleDown instead and pass the zoom of the widget to them.
```

If my assumption is correct, that's all that this PR does, right?

Small remark (for a future PR): "autoScaleUp" should probably be renamed to "scaleUp" in order to remain consistent with "scaleDown", which also accepts the zoom as a parameter.

That being said, since I don't expect any "breaking" changes based on my comments, I'm going to test your code for regressions :-)

**Comment by akoch-yatta** [comment]
> How about these calls still in the code? I assume the one in `Display` is still OK (please confirm/rebut) but shouldn't the other ones be removed (in future PRs)?
The call into Taskbar will be adapted in https://github.com/eclipse-platform/eclipse.platform.swt/pull/1269
The calls in the tests are used to store (and later restore) this attribute for now. We should evaluate removing that when we really merged all PR, because any call to DPIUtils without passing the zoom, will still implicitly use hte deviceZoom attribute.
The call in Display is adapted in the Image PR, I think so https://github.com/eclipse-platform/eclipse.platform.swt/pull/1236

**Comment by fedejeanne** [comment]
> > How about these calls still in the code? I assume the one in `Display` is still OK (please confirm/rebut) but shouldn't the other ones be removed (in future PRs)?
> 
> The call into Taskbar will be adapted in #1269. The calls in the tests are used to store (and later restore) this attribute for now. We should evaluate removing that when we really merged all PRs, because any call to DPIUtils without passing the zoom, will still implicitly use hte deviceZoom attribute. The call in Display is adapted in the Image PR, I think, so #1236

Great, thank you @akoch-yatta !

@amartya4256 you should only change the commit text then and add the extra assertion in the test and then this PR is good to go. My testing went well: I didn't find any regressions :-)

**Review by fedejeanne — APPROVED** [review]

**Comment by fedejeanne** [comment]
> As soon as the checks are green (specially `Tests / Build`) we can merge this PR. Thank you all for your input!

That one passed and "only" the `Jenkins` check failed with oh-so-annoying: `403 Killed` inside `script.sh.copy`. Last re-run, I'm merging afterwards.
